Media Query Mixins

AlecRaeside opened the pull request.

AlecRaeside committed:
demo files and refactored respond mixins

AlecRaeside committed:
working version

AlecRaeside committed:
bad attempt at range mixin refactor

AlecRaeside committed:
strip back to 2 flexible mixins

AlecRaeside committed:
cleanup

AlecRaeside committed:
updated breakpoint mixin demos

AlecRaeside committed:
try including css

AlecRaeside committed:
remove bad files

AlecRaeside committed:
no need for style.css

AlecRaeside committed:
remove local demo

AlecRaeside committed:
remove redundant ignore

chris-pearce commented on core/mixins/_media-queries.scss:
```diff
@@ -35,189 +39,39 @@
 /* Simple minimum and maximum width/height values
    ========================================================================== */
```
This type of comment is reserved for when you have a huge amount of CSS so you can easily break it up but most of the framework doesn't need them as we're very modular. So just the standard will do e.g. https://github.com/westfieldlabs/scally/blob/master/core/mixins/_convert-px-to-em-rem.scss#L31-L33.

chris-pearce replied:
I'll be updating the [CSS Guidelines](https://github.com/westfield/css_guidelines) once Scally is complete as we have some new authoring guidelines mainly around how we write comments.

chris-pearce commented on core/mixins/_media-queries.scss:
```diff
@@ -10,15 +10,19 @@
  * https://developer.mozilla.org/en-US/docs/Web/Guide/CSS/Media_queries.
```
Let's remove this. And add the `@demo` URL as per: https://github.com/westfieldlabs/scally/blob/master/core/mixins/_convert-px-to-em-rem.scss#L23-L24.

chris-pearce commented on core/mixins/_media-queries.scss:
```diff
@@ -35,189 +39,39 @@
-    @media (min-height: ($value / 16)+em) {
-      @content;
-    }
+@mixin respond-to($value, $axis: "width", $limit: "min") {
```
I'm thinking `$limit` before `$axis` as I reckon `$limit` will get used more? So we can write mq's like:

```
@include respond-to(lap, max) { ... }
```

chris-pearce replied:
Not sure about the name `$axis` as straight away I'm thinking `x`, `y`, and `z` as the values.

chris-pearce commented on core/mixins/_media-queries.scss:
```diff
@@ -10,15 +10,19 @@
+     .foo {background: red;}
+   }
+
+   @include respond-to(500, $axis: "height") {
```
Couldn't this be the less verbose version?

```
@include respond-to(500, height) {
```

But see comment below about the order of these arguments.

chris-pearce commented on core/mixins/_media-queries.scss:
```diff
@@ -35,189 +39,39 @@
+  // If value is a named breakpoint
+  @if type-of($value) == "string" {
+    $value: map-get($breakpoints, $value);
   }
```
Can we include `@warn`'s?

chris-pearce replied:
As it fails if you enter a string that isn't one of the breakpoints.

chris-pearce commented on core/mixins/_media-queries.scss:
```diff
@@ -35,189 +39,39 @@
-  @if $height {
-    @media (min-height: ($min-value / 16)+em) and (max-height: ($max-value / 16)+em) {
-      @content;
+    // If max-value is omitted, use the next named breakpoint after the min-value breakpoint
```
Not sure about this, I'm leaning towards to just rendering an error if `$max-vaue` is omitted?

AlecRaeside replied:
This is included to cover the functionality of the `respond-range-fixed` mixin in CC https://github.com/AlecRaeside/customer_console/blob/master/app/assets/stylesheets/partials/utils/_mixins.scss#L768 which is used in a couple of place in CC

chris-pearce replied:
Ah right

chris-pearce commented on core/mixins/_media-queries.scss:
```diff
@@ -35,189 +39,39 @@
+    // If max-value is omitted, use the next named breakpoint after the min-value breakpoint
+    @if $max-value == 0 {
+      $breakpoint-values: map-values($breakpoints);
+      $breakpoint-index:index($breakpoint-values, $min-value);
```
missing space after `$breakpoint-index:`.

AlecRaeside committed:
move logic into to-em function. improve the respond range mixin

AlecRaeside committed:
add space

AlecRaeside committed:
convert to spaces

AlecRaeside commented:
I've updated after code review. I've also moved logic to get the breakpoint values from breakpoint names into the to-em function. This has made the mixins much more DRY as the name validity checks are not repeated everywhere. @chris-pearce

chris-pearce commented:
@AlecRaeside nice work, looking very clean, just some authoring things, we're doing 2 line breaks now between the partial heading and intro block: https://github.com/AlecRaeside/scally/blob/refactor-mixins/core/mixins/_media-queries.scss#L4. And same spacing between the style blocks e.g. https://github.com/westfieldlabs/scally/blob/master/core/mixins/_gradients.scss. Also we're wrapping at 100 characters, I've got this in my Sublime Text preferences file:

```
"rulers":
 [
   100
 ],
```

This will all be documented properly in the new Guidelines once Scally is done.

AlecRaeside committed:
add spacing

AlecRaeside commented:
added spacing. I've added that config cheers.  line 66 i exactly 100 chars long

AlecRaeside committed:
wrap comment
